---
title: "Orthofinder"
author: "Soyoung Jung"
date: "4/21/2021"
output: html_document
---
##Install the orthofinder
```{bash}
conda install -c bioconda orthofinder
```

And I have the orthofinder installed:
```{bash}

Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.8.3
  latest version: 4.10.1

Please update conda by running

    $ conda update -n base -c defaults conda



# All requested packages already installed.
```

To check it was installed correctly:
```{bash}
orthofinder -h
```

And I was able to get:
```{bash}
OrthoFinder version 2.2.7 Copyright (C) 2014 David Emms

SIMPLE USAGE:
Run full OrthoFinder analysis on FASTA format proteomes in <dir>
  orthofinder [options] -f <dir>

Add new species in <dir1> to previous run in <dir2> and run new analysis
  orthofinder [options] -f <dir1> -b <dir2>

OPTIONS:
 -t <int>          Number of parallel sequence search threads [Default = 16]
 -a <int>          Number of parallel analysis threads [Default = 1]
 -M <txt>          Method for gene tree inference. Options 'dendroblast' & 'msa'
                   [Default = dendroblast]
 -S <txt>          Sequence search program [Default = blast]
                   Options: blast, mmseqs, blast_gz, diamond
 -A <txt>          MSA program, requires '-M msa' [Default = mafft]
                   Options: muscle, mafft
 -T <txt>          Tree inference method, requires '-M msa' [Default = fasttree]
                   Options: iqtree, raxml-ng, fasttree, raxml
 -s <file>         User-specified rooted species tree
 -I <int>          MCL inflation parameter [Default = 1.5]
 -x <file>         Info for outputting results in OrthoXML format
 -p <dir>          Write the temporary pickle files to <dir>
 -1                Only perform one-way sequence search 
 -n <txt>          Name to append to the results directory
 -h                Print this help text

WORKFLOW STOPPING OPTIONS:
 -op               Stop after preparing input files for BLAST
 -og               Stop after inferring orthogroups
 -os               Stop after writing sequence files for orthogroups
                   (requires '-M msa')
 -oa               Stop after inferring alignments for orthogroups
                   (requires '-M msa')
 -ot               Stop after inferring gene trees for orthogroups 

WORKFLOW RESTART COMMANDS:
 -b  <dir>         Start OrthoFinder from pre-computed BLAST results in <dir>
 -fg <dir>         Start OrthoFinder from pre-computed orthogroups in <dir>
 -ft <dir>         Start OrthoFinder from pre-computed gene trees in <dir>

LICENSE:
 Distributed under the GNU General Public License (GPLv3). See License.md

CITATION:
 When publishing work that uses OrthoFinder please cite:
 Emms D.M. & Kelly S. (2015), Genome Biology 16:157

 If you use the species tree in your work then please also cite:
 Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278
 Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914

```

So it seems like orthofinder is installed well.

And then, I run the orthofinder using: 
```{bash}
soyoungjung@Soyoungs-MacBook-Pro Bot563 % orthofinder -f SupplementalDataSet3
```

But then, this error massage came out: 
```{bash}
OrthoFinder version 2.2.7 Copyright (C) 2014 David Emms

2021-04-22 00:33:37 : Starting OrthoFinder
16 thread(s) for highly parallel tasks (BLAST searches etc.)
1 thread(s) for OrthoFinder algorithm

Checking required programs are installed
----------------------------------------
Test can run "makeblastdb -help" - ok
Test can run "blastp -help" - ok
Test can run "mcl -h" - ok
Test can run "fastme -i /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/SimpleTest.phy -o /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/SimpleTest.tre" - ok
ERROR: At least two species are required
ERROR: An error occurred, please review error messages for more information.
```

I think this is because I have all the amino acid sequences from all the different species in one fasta file.
So I made different fasta files based on the plant species in the same folder and then run the orthofinder again with 
```{bash}
orthofinder -f SupplementalDataSet3
```

And the result was like this:
```{bash}
OrthoFinder version 2.2.7 Copyright (C) 2014 David Emms

2021-04-22 09:33:50 : Starting OrthoFinder
16 thread(s) for highly parallel tasks (BLAST searches etc.)
1 thread(s) for OrthoFinder algorithm

Checking required programs are installed
----------------------------------------
Test can run "makeblastdb -help" - ok
Test can run "blastp -help" - ok
Test can run "mcl -h" - ok
Test can run "fastme -i /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/SimpleTest.phy -o /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/SimpleTest.tre" - ok

Dividing up work for BLAST for parallel processing
--------------------------------------------------
2021-04-22 09:33:51 : Creating Blast database 1 of 17
2021-04-22 09:33:51 : Creating Blast database 2 of 17
2021-04-22 09:33:51 : Creating Blast database 3 of 17
2021-04-22 09:33:51 : Creating Blast database 4 of 17
2021-04-22 09:33:51 : Creating Blast database 5 of 17
2021-04-22 09:33:51 : Creating Blast database 6 of 17
2021-04-22 09:33:52 : Creating Blast database 7 of 17
2021-04-22 09:33:52 : Creating Blast database 8 of 17
2021-04-22 09:33:52 : Creating Blast database 9 of 17
2021-04-22 09:33:52 : Creating Blast database 10 of 17
2021-04-22 09:33:52 : Creating Blast database 11 of 17
2021-04-22 09:33:53 : Creating Blast database 12 of 17
2021-04-22 09:33:53 : Creating Blast database 13 of 17
2021-04-22 09:33:53 : Creating Blast database 14 of 17
2021-04-22 09:33:53 : Creating Blast database 15 of 17
2021-04-22 09:33:53 : Creating Blast database 16 of 17
2021-04-22 09:33:54 : Creating Blast database 17 of 17

Running BLAST all-versus-all
----------------------------
Using 16 thread(s)
2021-04-22 09:33:54 : This may take some time....
2021-04-22 09:33:54 : Done 0 of 289
2021-04-22 09:33:59 : Done 100 of 289
2021-04-22 09:34:04 : Done 200 of 289
2021-04-22 09:34:08 : Done all-versus-all sequence search

Running OrthoFinder algorithm
-----------------------------
2021-04-22 09:34:08 : Initial processing of each species
2021-04-22 09:34:08 : Initial processing of species 0 complete
2021-04-22 09:34:09 : Initial processing of species 1 complete
2021-04-22 09:34:09 : Initial processing of species 2 complete
WARNING: Too few hits between species 3 and species 3 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 3 and species 6 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 3 and species 8 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 3 and species 11 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 3 and species 16 to normalise the scores, these hits will be ignored
2021-04-22 09:34:09 : Initial processing of species 3 complete
2021-04-22 09:34:09 : Initial processing of species 4 complete
2021-04-22 09:34:09 : Initial processing of species 5 complete
WARNING: Too few hits between species 6 and species 3 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 6 and species 6 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 6 and species 8 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 6 and species 11 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 6 and species 16 to normalise the scores, these hits will be ignored
2021-04-22 09:34:09 : Initial processing of species 6 complete
2021-04-22 09:34:09 : Initial processing of species 7 complete
WARNING: Too few hits between species 8 and species 3 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 8 and species 6 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 8 and species 8 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 8 and species 11 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 8 and species 16 to normalise the scores, these hits will be ignored
2021-04-22 09:34:09 : Initial processing of species 8 complete
2021-04-22 09:34:09 : Initial processing of species 9 complete
2021-04-22 09:34:09 : Initial processing of species 10 complete
WARNING: Too few hits between species 11 and species 3 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 11 and species 6 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 11 and species 8 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 11 and species 11 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 11 and species 16 to normalise the scores, these hits will be ignored
2021-04-22 09:34:09 : Initial processing of species 11 complete
2021-04-22 09:34:09 : Initial processing of species 12 complete
2021-04-22 09:34:09 : Initial processing of species 13 complete
2021-04-22 09:34:09 : Initial processing of species 14 complete
2021-04-22 09:34:09 : Initial processing of species 15 complete
WARNING: Too few hits between species 16 and species 3 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 16 and species 6 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 16 and species 8 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 16 and species 11 to normalise the scores, these hits will be ignored
WARNING: Too few hits between species 16 and species 16 to normalise the scores, these hits will be ignored
2021-04-22 09:34:09 : Initial processing of species 16 complete
2021-04-22 09:34:14 : Connected putatitive homologs
2021-04-22 09:34:14 : Written final scores for species 0 to graph file
2021-04-22 09:34:14 : Written final scores for species 1 to graph file
2021-04-22 09:34:15 : Written final scores for species 2 to graph file
2021-04-22 09:34:15 : Written final scores for species 3 to graph file
2021-04-22 09:34:15 : Written final scores for species 4 to graph file
2021-04-22 09:34:15 : Written final scores for species 5 to graph file
2021-04-22 09:34:15 : Written final scores for species 6 to graph file
2021-04-22 09:34:15 : Written final scores for species 7 to graph file
2021-04-22 09:34:15 : Written final scores for species 8 to graph file
2021-04-22 09:34:15 : Written final scores for species 9 to graph file
2021-04-22 09:34:15 : Written final scores for species 10 to graph file
2021-04-22 09:34:15 : Written final scores for species 11 to graph file
2021-04-22 09:34:15 : Written final scores for species 12 to graph file
2021-04-22 09:34:15 : Written final scores for species 13 to graph file
2021-04-22 09:34:15 : Written final scores for species 14 to graph file
2021-04-22 09:34:15 : Written final scores for species 15 to graph file
2021-04-22 09:34:15 : Written final scores for species 16 to graph file
2021-04-22 09:34:15 : Ran MCL

Writing orthogroups to file
---------------------------
Orthogroups have been written to tab-delimited files:
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthogroups.csv
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthogroups.txt (OrthoMCL format)
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthogroups_UnassignedGenes.csv
2021-04-22 09:34:15 : Done orthogroups

Analysing Orthogroups
=====================

Calculating gene distances
--------------------------
2021-04-22 09:34:18 : Done
Using fallback species tree inference method

Inferring gene and species trees
--------------------------------
2021-04-22 09:34:20 : Done 0 of 2

Best outgroup(s) for species tree
---------------------------------
2021-04-22 09:34:21 : Starting STRIDE
2021-04-22 09:34:21 : Done STRIDE
Observed 0 well-supported, non-terminal duplications. 0 support the best roots and 0 contradict them.
Best outgroups for species tree:
  Chlamydomonas_reinhardtii
  Volvox_carteri
  Spinacia_oleracea
  Solanum_lycopersicum
  Selaginella_moellendorffii
  Amborella_trichopoda
  Selaginella_moellendorffii, Volvox_carteri, Amborella_trichopoda, Chlamydomonas_reinhardtii, Marchantia_polymorpha, Mycobacterium_tuberculosis
  Selaginella_moellendorffii, Sorghum_bicolor, Volvox_carteri, Amborella_trichopoda, Chlamydomonas_reinhardtii, Marchantia_polymorpha, Oriza_sativa, Mycobacterium_tuberculosis
  Mycobacterium_tuberculosis, Chlamydomonas_reinhardtii, Volvox_carteri
  Sorghum_bicolor
  Oriza_sativa
  Arabidopsis_thaliana, Gossypium_raimondii, Brassica_rapa, Theobroma_cacao, Cucumis_sativus
  Mycobacterium_tuberculosis
  Theobroma_cacao
  Cucumis_sativus
  Brassica_rapa
  Arabidopsis_thaliana
  Medicago_truncatula
  Selaginella_moellendorffii, Mycobacterium_tuberculosis, Chlamydomonas_reinhardtii, Marchantia_polymorpha, Volvox_carteri
  Mycobacterium_tuberculosis, Chlamydomonas_reinhardtii, Marchantia_polymorpha, Volvox_carteri
  Arabidopsis_thaliana, Gossypium_raimondii, Brassica_rapa, Theobroma_cacao
  Populus_trichocarpa, Solanum_lycopersicum, Theobroma_cacao, Arabidopsis_thaliana, Brassica_rapa, Gossypium_raimondii, Cucumis_sativus, Medicago_truncatula
  Gossypium_raimondii, Theobroma_cacao
  Populus_trichocarpa
  Oriza_sativa, Sorghum_bicolor
  Arabidopsis_thaliana, Brassica_rapa
  Gossypium_raimondii
  Populus_trichocarpa, Medicago_truncatula
  Chlamydomonas_reinhardtii, Volvox_carteri
  Marchantia_polymorpha
  Populus_trichocarpa, Solanum_lycopersicum, Medicago_truncatula


Multiple potential species tree roots were identified, only one will be analyed.
================================================================================

Reconciling gene trees and species tree
---------------------------------------
Outgroup: Chlamydomonas_reinhardtii
2021-04-22 09:34:21 : Starting Recon and orthologues
2021-04-22 09:34:21 : Starting OF Orthologues
2021-04-22 09:34:21 : Done 0 of 1
2021-04-22 09:34:21 : Done OF Orthologues
2021-04-22 09:34:21 : Done Recon

Writing results files
=====================
2021-04-22 09:34:21 : Done orthologues
Orthogroups have been written to tab-delimited files:
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthogroups.csv
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthogroups.txt (OrthoMCL format)
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthogroups_UnassignedGenes.csv

Gene trees:
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthologues_Apr22/Gene_Trees/

Rooted species tree:
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthologues_Apr22/SpeciesTree_rooted.txt

Species-by-species orthologues:
   /Users/soyoungjung/Documents/Bot563/Bot563/SupplementalDataSet3/Results_Apr22_3/Orthologues_Apr22/Orthologues/

Orthogroup statistics:
   Statistics_PerSpecies.csv   Statistics_Overall.csv   Orthogroups_SpeciesOverlaps.csv

OrthoFinder assigned 45 genes (97.8% of total) to 1 orthogroups. Fifty percent of all genes were in orthogroups with 45 or more genes (G50 was 45) and were contained in the largest 1 orthogroups (O50 was 1). There were 1 orthogroups with all species present and 0 of these consisted entirely of single-copy genes.

CITATION:
 When publishing work that uses OrthoFinder please cite:
 Emms D.M. & Kelly S. (2015), Genome Biology 16:157

 If you use the species tree in your work then please also cite:
 Emms D.M. & Kelly S. (2017), MBE 34(12): 3267-3278
 Emms D.M. & Kelly S. (2018), bioRxiv https://doi.org/10.1101/267914
```

